test/HLSL/passes/multi_dim_one_dim/gep_addrspacecast_gep.ll

```diff
@@ -101,3 +102,97 @@ entry:
+  %gep2 = getelementptr inbounds [9 x float], [9 x float] addrspace(3)* %gep1, i32 0, i32 %idx1
+  %load = load float, float addrspace(3)* %gep2
+  ret void
+}
+
+; Test that we compute the correct index when the global has 3 levels of
+; nested arrays and an addrspacecast.
+; CHECK-LABEL: @merge_gep_multi_level_with_addrspace
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = mul i32 %1, 3
+; CHECK:  %3 = add i32 %idx2, %2
+; CHECK:  %4 = getelementptr [6912 x float], [6912 x float] addrspace(3)* @ArrayOfArrayOfArray.1dim, i32 0, i32 %3
+; CHECK:  %5 = addrspacecast float addrspace(3)* %4 to float*
+; CHECK:  load float, float* %5
+define void @merge_gep_multi_level_with_addrspace(i32 %idx0, i32 %idx1, i32 %idx2) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x [3 x float]]], [256 x [9 x [3 x float]]] addrspace(3)* @ArrayOfArrayOfArray, i32 0, i32 %idx0
+  %asc  = addrspacecast [9 x [3 x float]] addrspace(3)* %gep0 to [9 x [3 x float]]*
+  %gep1 = getelementptr inbounds [9 x [3 x float]], [9 x [3 x float]]* %asc, i32 0, i32 %idx1
+  %gep2 = getelementptr inbounds [3 x float], [3 x float]* %gep1, i32 0, i32 %idx2
+  %load = load float, float* %gep2
+  ret void
+}
+
+; Test that we compute the correct index when the global has 3 levels of
+; nested arrays.
+; CHECK-LABEL: @merge_gep_multi_level
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = mul i32 %1, 3
+; CHECK:  %3 = add i32 %idx2, %2
+; CHECK:  %4 = getelementptr [6912 x float], [6912 x float] addrspace(3)* @ArrayOfArrayOfArray.1dim, i32 0, i32 %3
+; CHECK:  load float, float addrspace(3)* %4
+define void @merge_gep_multi_level(i32 %idx0, i32 %idx1, i32 %idx2) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x [3 x float]]], [256 x [9 x [3 x float]]] addrspace(3)* @ArrayOfArrayOfArray, i32 0, i32 %idx0
+  %gep1 = getelementptr inbounds [9 x [3 x float]], [9 x [3 x float]] addrspace(3)* %gep0, i32 0, i32 %idx1
+  %gep2 = getelementptr inbounds [3 x float], [3 x float] addrspace(3)* %gep1, i32 0, i32 %idx2
+  %load = load float, float addrspace(3)* %gep2
+  ret void
+}
+
+; Test that we compute the correct index when the addrspacecast includes both a
+; change in address space and a change in the underlying type. I did not see
+; this pattern in IR generated from hlsl, but we can handle this case so I am
+; adding a test for it anyway.
+; CHECK-LABEL: addrspace_cast_new_type
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = getelementptr [2304 x float], [2304 x float] addrspace(3)* @ArrayOfArray.1dim, i32 0, i32 %1
+; CHECK:  %3 = addrspacecast float addrspace(3)* %2 to i32*
+; CHECK:  load i32, i32* %3
+define void @addrspace_cast_new_type(i32 %idx0, i32 %idx1) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]] addrspace(3)* @ArrayOfArray, i32 0, i32 %idx0
+  %asc  = addrspacecast [9 x float] addrspace(3)* %gep0 to [3 x i32]*
```

Comment:
Wait, you're allowed to change the type in an addrspacecast?  I thought that would require a bitcast.

Reply:
Yeah, it seems a bit bonkers to me but apparently it is allowed. But for addrspace cast from/to the same addrspace you have to use a bitcast it says:

https://llvm.org/docs/LangRef.html#addrspacecast-to-instruction

> It can be a no-op cast or a complex value modification, depending on the target and the address space pair. Pointer conversions within the same address space must be performed with the bitcast instruction.

I interpreted the "complex value modification" to mean it can change the type and the bitcode was accepted as valid when I tried it. 

I was unable to get that pattern from the hlsl I tried though.

Reply:
If the pointer type can be changed by an addrspacecast, then won't this code break down if the shape is changed?  If you change the structure through the addrspacecast, then try to move the addrspace cast to the end, it may invalidate the GEP indices that are being moved across the addrspacecast and concatenated into one GEP.

For example, imagine a multi-dim array is changed to a single-dim array in the addrspacecast, then a GEP with a single index indexes into that array.  You can't move the addrspacecast and merge GEPs, since you can't just copy the one index into the single-dim array to index the multi-dim array before the addrspacecast.  I think if there's a type change, and you want to move the addrspacecast to the end, you have to insert a bitcast to replace the addrspacecast for the type change, and you can't merge GEPs across the bitcast.  Either that, or you have to just leave the addrspacecast where it is, and it blocks GEP merging across that operation.

To illustrate with the MergeGEP comment:
```cpp
//      gep(addrspacecast(gep(p0, gep_args0) to p1*), gep_args1)
// into
//      addrspacecast(gep(p0, gep_args0+gep_args1) to p1*)
```
This is no longer valid if addrspacecast changes the type, because `gep_args1` in `gep_args0+gep_args1` is indexing a different type than `gep_args1` was originally.

Also, as a side note, `p1*` is not the same in these two expressions.

Reply:
Yeah, I think it will break in a more aggressive change of type. Should I change it to not merge the geps if the addrspacecast changes types instead of trying to support the simple case?

Reply:
I think, to be safe, we should stop the merge if the addrspacecast changes the type in between GEPs.

Reply:
Sounds good to me. I'll put up a new PR to do that.

Reply:
Note that in the case I was looking at originally (the multi-dim-to-one-dim pass) this will cause that pass to crash because it requires a single level merged gep for the accesses.

So for the test case in that PR we will be expecting a crash.

Reply:
In that case, was an addrspacecast changing the type in the middle of GEPs being merged?

Reply:
No, I never hit the change of type in an addrspacecast in real life hlsl code. Just in artificial IR test cases.

Reply:
Made the change to stop the merge here: https://github.com/microsoft/DirectXShaderCompiler/pull/5914